_posts/2025-03-24-Working-with-data.md

```diff
@@ -0,0 +1,407 @@
+
+## Putting it all together
+
+The BciPy data handling process involves loading raw data, extracting triggers, filtering the data, and processing it for further analysis or model training. The TriggerHandler and RawData classes are used to load and process data in Calibration and Copy Phrase modes. The data is organized in a hierarchical file structure, and users can filter and process the data using various methods and tools provided by BciPy. The processed data can be used for training machine learning models, analyzing EEG signals, or converting data to the BIDS format for sharing with other researchers.
+
+Here's an example of how to put it all together to visualize ERP data:
+
+```python
+
+import bcipy.acquisition.devices as devices
+from bcipy.config import (DEFAULT_DEVICE_SPEC_FILENAME,
+                          DEFAULT_PARAMETERS_FILENAME, RAW_DATA_FILENAME,
+                          TRIGGER_FILENAME)
+from bcipy.helpers.acquisition import analysis_channels
+from bcipy.io.load import (load_experimental_data, load_json_parameters,
+                           load_raw_data)
+from bcipy.core.triggers import TriggerType, trigger_decoder
+from bcipy.helpers.visualization import visualize_erp
+from bcipy.signal.process import get_default_transform
+
+if __name__ == '__main__':
+    import argparse
+    from pathlib import Path
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '-p',
+        '--path',
+        help='Path to the directory with raw_data to be converted',
+        required=False)
+    parser.add_argument("--show", dest="show", action="store_true")
+    parser.add_argument("--save", dest="save", action="store_true")
+    parser.set_defaults(show=False)
+    parser.set_defaults(save=False)
+    args = parser.parse_args()
+
+    path = args.path
+    if not path:
+        path = load_experimental_data(
+            message="Select the folder containing the raw_data.csv, parameters.json and triggers.txt",
+            strict=True
+        )
+
+    parameters = load_json_parameters(f'{path}/{DEFAULT_PARAMETERS_FILENAME}',
+                                      value_cast=True)
+
+    # extract all relevant parameters
+    trial_window = parameters.get("trial_window", (0, 0.5))
+    prestim_length = parameters.get("prestim_length")
+    trials_per_inquiry = parameters.get("stim_length")
+    # The task buffer length defines the min time between two inquiries
+    # We use half of that time here to buffer during transforms
+    buffer = int(parameters.get("task_buffer_length") / 2)
+    # get signal filtering information
+    downsample_rate = parameters.get("down_sampling_rate")
+    notch_filter = parameters.get("notch_filter_frequency")
+    filter_high = parameters.get("filter_high")
+    filter_low = parameters.get("filter_low")
+    filter_order = parameters.get("filter_order")
```

Comment:
It might be useful to include a section on using the parameters `instantiate` method to extract a subset of data as a named tuple. That would simplify this example a bit.